llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1777 @@
+}
+
+bool PartialExecuter::runOnModule( llvm::Module & module )
+{
+	using namespace llvm;
+	//TODO: classifyFunctions(module) ?? possibly to precompute where to recurse;
+	
+	ModuleData data;
+	moduleData = &data;
+	data.initFunctionData(module);
+
+	bool changed2 = true;
+	while (changed2)
+	{
+		changed2 = false;
+		for (Function& F : module)
+		{
+			bool R = processFunction(F);
+//			if (R)
+//				llvm::errs() << "processed\t" << F.getName() << "\n";
+			changed2 |= R;
+		}
+	}
+
+	bool changed = false;
+	for (Function& F : module)
+	{
+		//llvm::errs() << F.getName() << "\taaa\n";
+		changed |= runOnFunction(F);
+
+		verifyFunction(F);
+		//llvm::errs() << F << "\n";
+		for (BasicBlock& BB : F)
+		{
+			std::set<llvm::BasicBlock*> pred;
+			for (BasicBlock* bb : predecessors(&BB))
+				pred.insert(bb);
+
+			for (PHINode& phi : BB.phis())
+			{
+//				llvm::errs() << BB << "\n";
+				for (BasicBlock* bb : pred)
+					assert(phi.getBasicBlockIndex(bb) != -1);
+			}
+		}
+	}
+
+	return changed;
+}
+
+bool PartialExecuter::processFunction(llvm::Function& F)
+{
+	using namespace llvm;
+//For each SCC (process in order)
+//	if single point of entry: start from there, otherwise (no point of entry -> all unreachable / multiple ones -> all reachable)
+//	execute from single point of entry (+ state that's intersection of possible path to it)
+//	if (unconditional or decidable) -> follow path, otherwise
+//		if immediate dominator is in the SCC, jump there(a refinement would be immediate dominator in SCC while invalidating all otherwise reachable SCC) or bail out
+//	bail out:
+//		sign as reachable all out-going edges from the SCC (keeping only the state pre-SCC) + all BB in the SCC
+
+	if (F.isDeclaration())
+		return false;
+
+	FunctionData& data = moduleData->getFunctionData(F);
+
+	bool hasIndirectUse = false;
+
+	int X = 0;
+	for (const Use &U : F.uses())
+	{
+		const User *FU = U.getUser();
+		if (!isa<CallInst>(FU) && !isa<InvokeInst>(FU))
+		{
+			hasIndirectUse = true;
+			continue;
+		}
+		const CallBase* CS = cast<CallBase>(FU);
+		if (CS->isCallee(&U))
+		{
+			X++;
+			
+//			data.visitCallBase(CS);	
+		}
+		else
+		{
+			hasIndirectUse = true;
+		}
+	}
+	
+	
+	if (!hasIndirectUse && F.getLinkage() != GlobalValue::ExternalLinkage)
+		;
+	else
+	{
+		data.visitNoInfo();	
+	}
+
+	bool changed = false;
+	while (const auto* toBeVisited = data.getSomethingToVisit())
+	{
+		changed = true;
+		data.visitCallEquivalent(*toBeVisited);
+	}	
+	return changed;
+}
+
+bool PartialExecuter::runOnFunction(llvm::Function& F)
+{
+	using namespace llvm;
+
+	if (F.isDeclaration())
+		return false;
+
+	bool changed = false;
+
+	FunctionData& data = moduleData->getFunctionData(F);
+	if (data.hasModifications())
+	{
+			//TODO: Check no CE are in the globals we are loading from
+			//TODO: collect data on used Globals -> force alignment on those
+		data.emitStats();
+		data.cleanupBB();
+		changed = true;
+	}
+
+	if (!data.callEquivalentQueue.empty())
```

Comment:
IIUC this loop is used to replace constant args in the code itself, it should be handled by Os passes already (arg promotion)